### ISSUE TITLE
Port built-in graphs to Python

### ISSUE THREAD

**Issue by gcarleo** [netket/netket#371]
In the process it might be good to add wrappers for existing graph libraries (say, networkx)

**Comment by VolodyaCO** [comment]
Although networkx has good support, it is extremely slow, especially for large graphs. I recommend [graph-tool](https://graph-tool.skewed.de/) (I have used it for learning dynamics of contagion models, and it exponentially outperforms networkx)

**Comment by gcarleo** [comment]
Good suggestion @VolodyaCO thanks, I wasn't aware of this library. However it seems dependent on Boost and it's compiled C++. Does it come in a pip package?

**Comment by VolodyaCO** [comment]
When I installed it I had to do it with a [non-standard conda channel](https://medium.com/@ronie/installing-graph-tool-for-python-3-on-anaconda-3f76d9004979). Apparently it is now included in [conda-forge](https://anaconda.org/conda-forge/graph-tool). I'm not sure it's in a pip package...

**Comment by gcarleo** [comment]
Alright then it's a no-go for the moment. We can stick to igraph and networkx that have better support. For our tasks they should be relatively fast, since in most cases the lists of edges etc are computed only once.

**Comment by VolodyaCO** [comment]
Well, if it's for small graphs, then surely this is not a critical performance issue and the best decision is to stay with networkx, as it is the most used one (with the best support).

**Comment by VolodyaCO** [comment]
@gcarleo I could try to help, but I am unfamiliar with the C++ code. If you give me specific tasks I can try to do them. However, maybe a member of the developer team is able to perform this work quicker and better.

**Comment by gcarleo** [comment]
@VolodyaCO a specific first task to start with would be to write an ```AbstractGraph``` class in python (say, similar to [this style](https://github.com/netket/netket/blob/09e011fdbe20e85cf65b663e896d19aa942a7a85/netket/hilbert/abstract_hilbert.py) ) that defines what the common interface of the graphs in python should be. In practice, it should have the same methods of the graph classes we already have. 

The pull request should be done to the branch v3.0.

**Comment by VolodyaCO** [comment]
@gcarleo can I write the CustomGraph, and the rest of the graphs using networkx?

**Comment by gcarleo** [comment]
Yes that sounds good, basically ```HyperCube``` and ```CustomGraph``` are just special [cases](https://networkx.github.io/documentation/stable/reference/generated/networkx.generators.lattice.hypercube_graph.html#networkx.generators.lattice.hypercube_graph) of already existing networkx graphs . The only trickier case is ```Lattice```, but maybe we could use some other existing library, for example https://pypi.org/project/pbcpy/ or others

**Comment by VolodyaCO** [comment]
@gcarleo do you want me to check that the nodes are always labeled by integers? Or should I give freedom to label nodes as strings?, e.g. `nk.graph.CustomGraph([("cavity 1", "cavity 2", "color 1")])`

**Comment by gcarleo** [comment]
I think that at the moment we construct CustomGraph just using a list of (integer) edges... maybe I would leave it this way since it seems easier to handle? I also realized that we actually don't need ``networkx``` do implement CustomGraph, it is easier to just work with ```scipy.sparse.csgraph``` since it has already all the methods we need!

**Comment by VolodyaCO** [comment]
@gcarleo I already wrote the CustomGraph using networkx, but if you want me to I can wrap the scipy version.

**Comment by gcarleo** [comment]
It was just to avoid introducing the explicit dependence on networkx for the simple cases we are interested in... if it's not too much work I would write CustomGraph with the scipy version, and leave the wrapper you already wrote for networkx as another class/converter to ```netket.AbstractGraph``` (maybe call it ```netket.graph.NetworkX``` that can be constructed from a networkx graph?)... up to you

**Comment by gcarleo** [comment]
But if you have already done it in networkx it's OK no problem... no need to rewrite it now

**Comment by gcarleo** [comment]
Yes sorry! I realize that there is a few methods missing in the scipy interface anyways... (is_bipartite is not there and finding automorphisms is definitely not there) so indeed it makes more sense to proceed with networkx...

**Comment by gcarleo** [comment]
It's amazing that now newtorkx has also a way to find automorphisms, before we had to use python-igraph... 

```python
graph=nx.Graph([[0,1],[1,2],[2,3],[3,0]])

ismags=nx.isomorphism.ISMAGS(graph,graph)
automorphisms=[[iso[i] for i in range(graph.size())] for iso in ismags.isomorphisms_iter(symmetry=False)]
```

**Comment by VolodyaCO** [comment]
I was checking that before as well... but I was a bit unsure of including something of the sort because in the C++ implementation the automorphisms are an input to the CustomGraph. However, it should be something that gets computed always, right @gcarleo ? i.e. it shouldn't be an argument of the init method.

**Comment by gcarleo** [comment]
Indeed, the old C++ implementation was quite limited in this. Now we can make it automatically computed when one calls ```.automorphisms()``` (I wouldn't compute it by default though, but only if one calls that method and wants the automorphisms)... 

Also, it would be a good idea to call the main class you are writing just ```netket.Graph```, it doesn't really make sense to call this ```CustomGraph``` anymore! (that was a kind of weird name that was left from version 1.0..)

**Comment by VolodyaCO** [comment]
Good, I'll have a couple of free hours tomorrow. I will try to finish the CustomGraph, and to add the Hypercube. Is the Edgeless graph necessary?

**Comment by VolodyaCO** [comment]
@gcarleo I have created the customgraph and hypercube classes with networkx. Shall I open a PR with these? Or do I only do it until I implement the Lattice graph?

**Comment by gcarleo** [comment]
Yes you can open this for the moment, then we will see what do with Lattice thanks

**Comment by VolodyaCO** [comment]
@gcarleo I was checking the library pbcpy and it only receives 3d lattice vectors, do we want to have n dimensional lattices?

**Comment by gcarleo** [comment]
something that works in 1,2 and 3d is more than enough! 
I would suggest to first finish the simpler graphs now and then open a new PR on the ```Lattice``` class only? I think that we don't need to import or depend from pbcpy, we just need a function that computes distances with the minimum image convention, if periodic boundary conditions are used... see [for example ](https://gist.github.com/tovrstra/3566972)

**Comment by gcarleo** [comment]
Here is a basic implementation of the methods needed 
```python
import itertools
import numpy as np

basis_vectors = [[1, 0], [0, 1]]
extent = [2, 2]
atoms_coord = [[0, 0], [0.5, 0.5]]


ranges = tuple([list(range(ex)) for ex in extent])
n_atoms = len(atoms_coord)

coord_to_site = []
site_to_coord = {}
atom_label = []

for r in itertools.product(*ranges):
    cell_coord = np.matmul(basis_vectors, r)

    for atom in atoms_coord:
        coord = np.asarray(atom) + cell_coord
        site_to_coord[tuple(coord)] = len(coord_to_site)
        coord_to_site.append(coord)

    atom_label += range(n_atoms)


n_sites = len(coord_to_site)

print(coord_to_site[5])
print(site_to_coord[(1.5, 0.5)])
print(atom_label[5])
```

The main remaining issue is how to compute distances with PBC using the coordinates of two atoms, as given in this code...

**Comment by VolodyaCO** [comment]
As I understand, the only remaining issue with the graphs is the automorphisms ordering. I added a comment about that in #405

**Comment by VolodyaCO** [comment]
I'll do the Lattice PR this weekend

**Comment by VolodyaCO** [comment]
@gcarleo just to be sure, the distance between two adjacent atoms is the euclidean distance between them, right? It is not 1 (1 leapfrog to reach the other atom).

**Comment by VolodyaCO** [comment]
I've checked some examples with the current `Lattice`, and it seems that edges are only formed for first-neighbours. Moreover, the distances are being computed as graph distances, and not euclidean distances:
```
>>> g=nk.graph.Lattice(basis_vectors=[[1,0],[0,1]],extent=[1,2], atoms_coord=[[0.0, 0.0], [0.5, 0.5], [1, 1]], pbc=[False, False])
>>> g.distances
[[0, 2, 1, 3, 2, 4],
 [2, 0, 1, 1, 2, 2],
 [1, 1, 0, 2, 1, 3],
 [3, 1, 2, 0, 1, 1],
 [2, 2, 1, 1, 0, 2],
 [4, 2, 3, 1, 2, 0]]
```

However, moving the central atom seems to break things:
```
>>> g=nk.graph.Lattice(basis_vectors=[[1,0],[0,1]],extent=[1,2], atoms_coord=[[0.0, 0.0], [0.6, 0.6], [1, 1]], pbc=[False, False])
>>> g.distances
[[0, -1, -1, -1, -1, -1],
 [-1, 0, -1, -1, -1, -1],
 [-1, -1, 0, -1, 1, -1],
 [-1, -1, -1, 0, -1, 1],
 [-1, -1, 1, -1, 0, -1],
 [-1, -1, -1, 1, -1, 0]]
```

**Comment by gcarleo** [comment]
Maybe @martamau can help understand that? Thanks!
(@martamau , we are trying to move your Lattice code to pure python)

**Comment by gcarleo** [comment]
General design goal should be the distances returned by the distances methods are indeed graph distances (i.e. integers). There might (should) be also another method returning cartesian distances (or distances should accept an optional argument type="cartesian", or something)

In general, it makes sense to define graph distances in the sense that all cartesian distances on the lattice belong to a fixed set of distances (say (1,sqrt(2),2,etc) for a square lattice). Thus in this sense graph distance 1 means cartesian 1, graph distance 2 means sqrt(2) etc.

**Comment by gcarleo** [comment]
Unless I am mistaken, fixing the nearest neighbors (in cartesian distance) of a given node of the lattice should be enough to consistently determine all the other graph distances

**Comment by VolodyaCO** [comment]
I am a bit confused. I would do a lattice with Bravais primitive vectors as the Lattice generator. Each to contiguous nodes of the Bravais lattice would be joined by an edge. However, within each unit cell, there are atoms placed relative to the cell's origin. To me, it is not quite straightforward how to connect the atoms by edges. The first idea was to use the first neighbours of each atom. However, if I have atoms like in the example:

`>>> g=nk.graph.Lattice(basis_vectors=[[1,0],[0,1]],extent=[1,4], atoms_coord=[[0.0, 0.0], [0.6, 0.6]], pbc=[False, False])`

Then I'd have something like this:
![image](https://user-images.githubusercontent.com/31494271/82157652-6d79fd80-9848-11ea-8ebc-557667a84085.png)

where each atom is linked to its first neighbours. If the user states which atoms are connected by edges, then the graph could be replicated for the extent of the lattice.

**Comment by gcarleo** [comment]
@VolodyaCO the code I wrote here (a week ago I think) should be enough to define the coordinates of each atom (including the case when you have more than one atom per unit cell). 

The thing I didn't put in that code is how to find the neighbors of a given atom. 
The simple (but horribly slow) solution is just for each atom to loop through all the others and find the one that has the minimum distance (cartesian distance, with the minimum image convention). 
It is possible to make it faster, since you really don't need to loop through all the possible atoms...

**Comment by VolodyaCO** [comment]
I've been taking a look at the minimum image convention discussion, and it seems that for a general vector basis some intensive calculation might be needed. I am checking some alternatives such as periodic KDTree [[1](https://github.com/patvarilly/periodic_kdtree)], [[2](https://www.mdanalysis.org/mdanalysis/documentation_pages/lib/pkdtree.html)], [[3](https://docs.scikit-nano.org/dev/api/sknano.core.analysis.PeriodicCKDTree.html)], [[4](http://prody.csb.pitt.edu/_modules/prody/kdtree/kdtree.html)]

**Comment by gcarleo** [comment]
Yes this would be a very good solution (your [1]) but it seems it's an old python 2 code, and I am not sure it has been ported to modern scipy?

**Comment by gcarleo** [comment]
However the [scipy version](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.html) seems to have a boxsize option that takes into account PBCs. I am not sure they use the minimum image convention, but it might worth trying it

**Comment by VolodyaCO** [comment]
Yes, however, I am afraid that with skewed unit cells these implementations do not work. But I've worked with KDTree before, and it is fast!

**Comment by gcarleo** [comment]
I see... well, let's just write an implementation that works with non-skewed unit cells for the moment! We can just return an error message for the skewed case

**Comment by VolodyaCO** [comment]
Ok, I've done it for orthorhombic cells. I will clean the code and create the PR tonight.

**Comment by gcarleo** [comment]
Great thanks. 

Related, do you know why this fails with your Graph? 

```
    for i in range(5, 10):
        for j in range(i + 1, i * i):
            x = nx.dense_gnm_random_graph(i, j)
            y = Graph(list(x.edges()))
            assert y.is_connected() and nx.is_connected(x)
```

**Comment by gcarleo** [comment]
Also this fails
```python
    for i in range(1, 10):
        for j in range(1, i * i):
            x = nx.dense_gnm_random_graph(i, j)
            y = Graph(list(x.edges()))
            assert y.is_bipartite() == nx.is_bipartite(x)
```

**Comment by gcarleo** [comment]
Also 
```python
    edges = [(0, 1, 0), (1, 2, 1), (2, 3, 0), (0, 3, 1)]
    g = Graph(edges)

    print(g.edges(color=True))
```

prints an empty list (even without color=True)

**Comment by VolodyaCO** [comment]
Ok, I'll open the PR and I'll check those issues tomorrow.

**Comment by VolodyaCO** [comment]
@gcarleo maybe you want to change the order of the `Graph` arguments, as the first argument is `nodes` and the second argument is `edges`. You are passing the edges as `nodes`. Also, in general one can pass any object as a node. The convention is to pass integers, but this is not currently enforced. Here's what I found with your first example:

```
flag = False
for i in range(5, 10):
    for j in range(i + 1, i * i):
        x = nx.dense_gnm_random_graph(i, j)
        y = Graph(edges=list(x.edges()))
        if not (y.is_connected and nx.is_connected(x)):
            nx.draw(x, with_labels=True)
            plt.show()
            nx.draw(y.graph, with_labels=True)
            plt.show()
            print(x.edges)
            print(x.nodes)
            print(y.edges())
            print(y.graph.nodes)
            flag = True
            break
    if flag:
        break
```
(how do you add nice Python colors to these markdown codes?)
This code draws these graphs:
![image](https://user-images.githubusercontent.com/31494271/82456074-5bcb6c80-9a79-11ea-873f-92f95a2670e1.png)

Now, there is an error in the code:
`y = Graph(nodes=list(x.nodes()), edges=list(x.edges()))
`
should fix your test, but I am checking that in the `Graph` function, these lines
```
if nodes:
    node_names = sorted(nodes)
if edges:
    node_names = sorted(set((node for edge in edges_array for node in edge)))
```
should be changed to:
```
if nodes:
    node_names = sorted(nodes)
elif edges:
    node_names = sorted(set((node for edge in edges_array for node in edge)))
```
If `edges` are provided, the first code ignores every not connected node. The second one does not.

**Comment by gcarleo** [comment]
Thanks I have fixed these things in #419
